ext/MitsubaLoader/CMitsubaLoader.cpp

```diff
@@ -1,165 +1,408 @@
+    float phys_pg_tex_sz_rcp[_IRR_VT_MAX_PAGE_TABLE_LAYERS];
+    uint layer_to_sampler_ix[_IRR_VT_MAX_PAGE_TABLE_LAYERS];
+} precomputed;
+
+layout (set = 0, binding = 3, std430) restrict readonly buffer INSTR_BUF
+{
+	instr_t data[];
+} instr_buf;
+layout (set = 0, binding = 4, std430) restrict readonly buffer BSDF_BUF
+{
+	bsdf_data_t data[];
+} bsdf_buf;
+
+uint irr_glsl_VT_layer2pid(in uint layer)
+{
+    return precomputed.layer_to_sampler_ix[layer];
+}
+uint irr_glsl_VT_getPgTabSzLog2()
+{
+    return precomputed.pgtab_sz_log2;
+}
+float irr_glsl_VT_getPhysPgTexSzRcp(in uint layer)
+{
+    return precomputed.phys_pg_tex_sz_rcp[layer];
+}
+float irr_glsl_VT_getVTexSzRcp()
+{
+    return precomputed.vtex_sz_rcp;
+}
+#define _IRR_USER_PROVIDED_VIRTUAL_TEXTURING_FUNCTIONS_
+
+#include <irr/builtin/glsl/virtual_texturing/functions.glsl/7/8>
+
+#include <irr/builtin/glsl/utils/common.glsl>
+
+layout (set = 1, binding = 0, row_major, std140) uniform UBO {
+    irr_glsl_SBasicViewParameters params;
+} CamData;
+
+struct InstanceData
+{
+	mat4x3 tform;
+	vec3 normalMatrixRow0;
+	uint bsdf_instrOffset;
+	vec3 normalMatrixRow1;
+	uint bsdf_instrCount;
+	vec3 normalMatrixRow2;
+	uint _padding;//not needed
+	uvec2 prefetch_instrStream;
+	uvec2 nprecomp_instrStream;
+	uvec2 genchoice_instrStream;
+	uvec2 emissive;
+};
+layout (set = 0, binding = 5, row_major, std430) readonly restrict buffer InstDataBuffer {
+	InstanceData data[];
+} InstData;
+
+vec3 irr_glsl_MC_getCamPos()
+{
+	vec3 campos = irr_glsl_SBasicViewParameters_GetEyePos(CamData.params.NormalMatAndEyePos);
+	return campos;
+}
+instr_t irr_glsl_MC_fetchInstr(in uint ix)
+{
+	return instr_buf.data[ix];
+}
+bsdf_data_t irr_glsl_MC_fetchBSDFData(in uint ix)
+{
+	return bsdf_buf.data[ix];
+}
+#define _IRR_USER_PROVIDED_MATERIAL_COMPILER_GLSL_BACKEND_FUNCTIONS_
+)";
+_IRR_STATIC_INLINE_CONSTEXPR const char* FRAGMENT_SHADER_IMPL = R"(
+#include <irr/builtin/glsl/format/decode.glsl>
+
+instr_stream_t getEvalStream()
+{
+	instr_stream_t stream;
+	stream.offset = InstData.data[InstanceIndex].bsdf_instrOffset;
+	stream.count = InstData.data[InstanceIndex].bsdf_instrCount;
 
-	return rgb19e7;
+	return stream;
 }
+//rem'n'pdf and eval use the same instruction stream
+instr_stream_t getRemAndPdfStream()
+{
+	return getEvalStream();
+}
+instr_stream_t getGenChoiceStream()
+{
+	instr_stream_t stream;
+	uvec2 s = InstData.data[InstanceIndex].genchoice_instrStream;
+	stream.offset = s.x;
+	stream.count =  s.y;
+
+	return stream;
+}
+instr_stream_t getTexPrefetchStream()
+{
+	uvec2 s = InstData.data[InstanceIndex].prefetch_instrStream;
+	instr_stream_t stream;
+	stream.offset = s.x;
+	stream.count = s.y;
 
-static bsdf::SBSDFUnion bsdfNode2bsdfStruct(const CElementBSDF* _node)
+	return stream;
+}
+instr_stream_t getNormalPrecompStream()
+{
+	uvec2 s = InstData.data[InstanceIndex].nprecomp_instrStream;
+	instr_stream_t stream;
+	stream.offset = s.x;
+	stream.count = s.y;
+
+	return stream;
+}
```

Comment:
we could use a single offset and keep the streams together in memory, then work out the other offsets from stream counts.

Reply:
same as https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r509236731